### ISSUE TITLE
[BUG] npm creates its cache directory in my $HOME

### ISSUE THREAD

**Issue by WhyNotHugo** [npm/cli#1451]
# What / Why

`npm` keeps creating its cache directory in my `$HOME`. I keep very few directories in my `$HOME` to keep things findable, and its very annoying when apps start littering stuff there.

## When

Any time I run things like `npm install blah`.

## Where

```console
$ uname -srmo
Linux 5.7.2-arch1-1 x86_64 GNU/Linux
```

### Current Behavior

A directory is creted in `$HOME` and cache data stored in there.

### Steps to Reproduce

0. Run `rm -rf .cache` to delete your existing cache.
1. Navigate to a directory with an existing `package.json`.
2. Run `npm install`.

### Expected Behavior

The cache directory should be in [`$XDG_CACHE_HOME`](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) on Linux. Specifically, it should be `$XDG_CACHE_HOME/npm`. `$XDG_CACHE_HOME` should default to `.cache` if undefined. 

I believe on macOS there's a different spec for default locations, and things like cache are stored in a different place. I don't know the name of the spec so can't easily link in here.

## Who

* n/a

## References

[XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html)

**Comment by WhyNotHugo** [comment]
Oh, there's an [implementation for nodejs](https://github.com/sindresorhus/xdg-basedir)!

**Comment by zangornjak** [comment]
Duplicate of https://github.com/npm/cli/issues/1305. It started around the time I created the issue, but I don't remember which version introduced this bug exactly.

**Comment by WhyNotHugo** [comment]
Not a duplicate. This bug is about the default behaviour being broken (i.e.: placing cache outside of where all application's cache should be).

#1305 is about the behaviour when overriding the default _also_ being broken.

**Comment by zangornjak** [comment]
I see, I hope you are able persuade them to implement this so the environment flag can be dropped in favor of a sane default.

**Comment by zangornjak** [comment]
I've updated the comment above with the previous discussion on this subject.

Reading through that (opened in late 2014) I get the feeling that it's best to try other solutions to avoid home pollution in the meantime. I've tried the approach described in [ArchWiki - XDG_Base_Directory](https://wiki.archlinux.org/index.php/XDG_Base_Directory) and it works while also solving the default overriding being broken.

Coincidentally it's the "use the config" approach being pointed at by the locker of the issue above.

**Comment by astier** [comment]
npm is one of the few programs which keeps littering my $HOME. Annoying.

**Comment by lizzielizzielizzie** [comment]
tl;dr: If you are seeing this when running `npm ci`, you may need to upgrade npm to at least version `6.11.0`

-------------

I've been digging into this issue (or rather, [this](https://github.com/getsentry/sentry-cli/issues/856) issue) for some time now. In my case, a package was downloading an additional binary to the directory specified by `process.env.npm_config_cache`, and if unset, falling back to `~/.npm`.  The thing was, I *had* specified a cache directory. So I did some further digging:

Given the `following package.json`,
```json
{
  "scripts": {
    "install": "node --eval 'console.log(process.env.node_config_cache)'"
  }
}
```

Running the following:
- `npm run-script install`: variable was set (**note:** this runs the script as a script, *not the lifecycle hook*)
- `npm install`: variable was set (I wasn't expecting this!)
- `npm ci`: variable was **not** set, even though it was run as part of the lifecycle hooks

That lead me to discover [this](https://github.com/nodejs/node/issues/28987) regression. After upgrading to the latest version of `npm` I no longer have issues with the cache appearing in unexpected places.

**Comment by txtsd** [comment]
What's the holdup on this? Should be an easy fix to change and prefer a different directory imo.

**Comment by darcyclarke** [comment]
npm `v6` is no longer in active development; We will continue to push security releases to `v6` at our team's discretion as-per our [Support Policy](https://github.com/npm/cli/wiki/Support-Policy).

If your bug is reproducible on `v7`, please re-file this issue using our [new issue template](https://github.com/npm/cli/issues/new?assignees=&labels=Bug%2CNeeds+Triage%2CRelease+7.x&template=bug_7.yml&title=%5BBUG%5D+%3Ctitle%3E).

If your issue was a feature request, please consider opening a new [RRFC](https://github.com/npm/rfcs/issues) or [RFC](https://github.com/npm/rfcs). If your issue was a question or other idea that was not CLI-specific, consider opening a discussion on our [feedback repo](https://github.com/npm/feedback)

**Closing:** This is an automated message.

**Comment by WhyNotHugo** [comment]
Discussion on this issue has now moved here:  https://github.com/npm/rfcs/issues/389
